File: src/main/java/net/masterthought/dlanguage/utils/DResolveUtil.java
package net.masterthought.dlanguage.utils;

import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiElement;
import com.intellij.psi.search.GlobalSearchScope;
import net.masterthought.dlanguage.psi.*;
import net.masterthought.dlanguage.psi.interfaces.DNamedElement;
import net.masterthought.dlanguage.psi.interfaces.containers.ConstructorContainer;
import net.masterthought.dlanguage.psi.interfaces.containers.Container;
import net.masterthought.dlanguage.psi.interfaces.containers.ContainerUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static com.intellij.psi.util.PsiTreeUtil.findChildOfType;
import static com.intellij.psi.util.PsiTreeUtil.getParentOfType;
import static java.util.Collections.EMPTY_SET;
import static net.masterthought.dlanguage.index.DModuleIndex.getFilesByModuleName;
import static net.masterthought.dlanguage.psi.interfaces.containers.ContainerUtil.getContainedDeclarationsWithPlaceHolders;
import static net.masterthought.dlanguage.utils.PlaceHolderUtils.fillPlaceHolders;

/**
 * so instead of doing some recursive complicated garbage this should work as follows:
 * getDeclarationsVisibleFromElement, collects all declarations at the element in question, and whenever it hits upon
 * inheritance or a mixin, it puts a placeholder in its place and continues. When done, it loops over and fills in the placeholders.
 * for mixins, the mixinable should already be in the collection of declarations, so simple take the contents of the mixinable
 * and add its declarations. For inherited classes the inherited symbols are first represented by a placeholder. The placeholder
 *  is filled by first checking if a matching declaration has already been found, and if not resolving one.
 *
 *
 * todo when constructor resolution fails, the class/struct should be resolved instead
 * todo match arguments to resolve overloaded functions
 * todo improve performance somehow -- stub index/file index for declarations?
 * todo create much more expansive unittests
 * todo when returning a result resolve the identifier not the entire declaration, however this may mess with some internal code in this file and in the ContainerUtil file
 * todo make induvidual symbol imports work
 * Created by franc on 1/18/2017.
 */
public class DResolveUtil {
    static Logger log = Logger.getInstance(DResolveUtil.class);

    public static
    @NotNull
    Set<? extends DNamedElement> findDefinitionNodes(@NotNull DNamedElement element) {
        if (!(element instanceof DLanguageIdentifier))
            return EMPTY_SET;//prevent resolving definitions
        return findDefinitionNodes((DLanguageFile) element.getContainingFile(), (DLanguageIdentifier) element);

    }

    private static
    @NotNull
    Set<? extends DNamedElement> findDefinitionNodes(DLanguageFile dLanguageFile, DLanguageIdentifier element) {
        //seven major possibilities:
        //1. identifier is part of import statement, and should resolve to a file
        //2. identifier is a normal function/class name/etc..
        //  2.a.overloaded functions, specifically resolving the correct overloaded functions
        //3. identifier is a constructor/destructor usage
        //  3.a.overloaded constructors, specifically resolving the correct overloaded functions
        //4. resolving something referred to with its full name eg. foo.bar.something();
        //5. resolving a member function/var of a class/struct/template etc...
        //  ex: Class a =new Class(); a.foo(); //resolving foo
        final PsiElement parent = element.getParent();
        //identifiers within templates break this approach
        if (!dLanguageFile.getText().contains(element.getText()) &&
            getParentOfType(parent, DLanguageImportDeclaration.class) == null)
            return Collections.emptySet();//optimization to prevent unnecessary searching
        if (getParentOfType(parent, DLanguageImportDeclaration.class) != null) {
            //#1 is true
            return findModuleDefinitionNodes(getParentOfType(element, DLanguageModuleFullyQualifiedName.class));
        }
        if (parent.getText().contains(".") && !parent.getText().contains("{") | parent instanceof DLanguageModuleFullyQualifiedName
//            parent instanceof DLanguageUnaryExpression ||
//            parent instanceof DLanguageIdentifierList ||
//            parent instanceof DLanguagePostfixExpression ||
//            parent instanceof DLanguageModuleFullyQualifiedName ||
//            parent instanceof DLanguagePrimaryExpression ||
//            parent instanceof DLanguageDotIdentifier ||
//            parent instanceof DLanguageSymbol ||
//            parent instanceof DLanguageSymbolTail ||
//            parent instanceof DLanguageMixinTemplateName ||
//            parent instanceof DLanguageQualifiedIdentifierList
            ) {
            //#4 or #5 are true
            //to distinguish between #4 and #5 resolve the rightmost identifier. eg. in foo.bar.gh(), find foo and resolve it
            return findDefinitionNodesStandard(element);
            //the identifier is within one of the following:
            // IdentifierList
            // UnaryExpression
            // PostfixExpression
            // PrimaryExpression
            // DotIdentifier
            // Symbol
            // SymbolTail
            // MixinTemplateName
            // QualifiedIdentifierList

        } else {
            //#2 or 3
            PsiElement current = parent;
            while (true) {
                if (current instanceof DLanguageNewExpression || current instanceof DLanguageNewExpressionWithArgs) {
                    //#3 is true
                    return findConstructorDefinitionNodes(element, current);
                }
                if (current instanceof DLanguageDeleteExpression) {
                    break;//todo resolving destructors?
                }
                if (current instanceof DLanguageCastExpression) {
                    //#2 is true
                    break;
                    // specifically a class,alias or struct
                }
                if (current instanceof DLanguageUnaryExpression) {
                    //#2 is true
                    break;
                }
                if (current instanceof DLanguageTypeidExpression) {
                    //#2 is true
                    break;
                }
                if (current instanceof DLanguageIsExpression) {
                    //#2 is true
                    break;
                }
                if (current instanceof DLanguageTraitsArgument) {
                    //#2 is true
                    break;
                }
                if (current == null)
                    break;
                current = current.getParent();
            }
            //default to #2:
            return findDefinitionNodesStandard(element);
        }
    }

    private static
    @NotNull
    Set<DNamedElement> findConstructorDefinitionNodes(@NotNull DLanguageIdentifier element, @NotNull PsiElement newExpression) {
        Set<DNamedElement> res = new HashSet<>();
        if (newExpression instanceof DLanguageNewExpression || newExpression instanceof DLanguageNewExpressionWithArgs) {
            for (DNamedElement dNamedElement : findDefinitionNodesStandard(element)) {
                if (dNamedElement instanceof ConstructorContainer) {
                    final Set<DNamedElement> memberDeclarations = fillPlaceHolders(ContainerUtil.getContainedDeclarationsWithPlaceHolders((Container) dNamedElement));
                    for (DNamedElement memberDeclaration : memberDeclarations) {
                        if (memberDeclaration instanceof DLanguageConstructor)
                            res.add(memberDeclaration);
                    }

                }
            }
        } else {
            throw new IllegalArgumentException("new expression must contain a new expression");
        }
        return res;
    }

    /**
     * for resolving anything that isn't a constructor and doesn't have a longform name like foo.bar.gh()
     *
     * @param element
     * @return
     */
    private static
    @NotNull
    Set<DNamedElement> findDefinitionNodesStandard(DLanguageIdentifier element) {
        Set<DNamedElement> res = new HashSet<>();
        if (element.getParent() instanceof DLanguageModuleFullyQualifiedName) {
            return EMPTY_SET;
        }
        for (DNamedElement dNamedElement : getDeclarationsVisibleFromElement(element, element.getParentContainer())) {
            if (dNamedElement.getName() != null && longNamesAreReferringToSameThing(dNamedElement.getFullName(), element.getName()) && !(dNamedElement instanceof DLanguageModuleDeclaration) && dNamedElement.getName().equals(element.getName())) {
                res.add(dNamedElement);
            }
        }
        return res;
    }

    /**
     * this method does most of the work for resolve and for completion.
     * it finds all declarations visible from an element
     * @param element         element being resolved if relevant. The element from which declarations are visible
     * @return declarations visible from element
     */
    @NotNull
    public static Set<DNamedElement> getDeclarationsVisibleFromElement(@NotNull DLanguageIdentifier element) {
        return getDeclarationsVisibleFromElement(element, element.getParentContainer());
    }

    @NotNull
    private static Set<DNamedElement> getDeclarationsVisibleFromElement(@Nullable DLanguageIdentifier element, Container parentContainer) {
        final Set<DNamedElement> withPlaceHolders = getVisibleElementsWithPlaceHolders(element, parentContainer);
        return fillPlaceHolders(withPlaceHolders);
    }

    @NotNull
    static Set<DNamedElement> getVisibleElementsWithPlaceHolders(DLanguageIdentifier element, Container parentContainer) {
        Set<DNamedElement> declarations = new HashSet<>();

        if (parentContainer instanceof DLanguageFile) {
            declarations.addAll(getContainedDeclarationsWithPlaceHolders(parentContainer));
            return declarations;//don't keep searching up after leaving the file
        }
        //add the declarations from here
        declarations.addAll(getContainedDeclarationsWithPlaceHolders(parentContainer));
        declarations.addAll(getVisibleElementsWithPlaceHolders(element, parentContainer.getParentContainer()));
        return declarations;
    }

    public static Set<DNamedElement> findModuleDefinitionNodes(DLanguageModuleFullyQualifiedName module) {
        Set<DNamedElement> res = new HashSet<>();
        for (DLanguageFile dLanguageFile : fromModulesToFiles(module.getProject(), Collections.singleton(module.getText()))) {
            if (findChildOfType(dLanguageFile, DLanguageModuleDeclaration.class) != null) {
                res.add(findChildOfType(dLanguageFile, DLanguageModuleDeclaration.class));
            } else {
                res.add(dLanguageFile);
            }
        }
        return res;
    }

    public static Set<DLanguageFile> fromModulesToFiles(Project project, Set<String> modules) {
        Set<DLanguageFile> filesFound = new HashSet<>();
        for (String module : modules) {
            List<DLanguageFile> files = getFilesByModuleName(project, module, GlobalSearchScope.allScope(project));
            filesFound.addAll(files);
        }
        return filesFound;
    }

    public static boolean longNamesAreReferringToSameThing(String longName, String shorterName) {
        final int i = longName.lastIndexOf(shorterName);
        if (i < 0)
            return false;
        return i == longName.length() - shorterName.length();
    }
}

File: src/main/java/net/masterthought/dlanguage/utils/DResolveUtil.kt
package net.masterthought.dlanguage.utils

import com.intellij.openapi.project.Project
import com.intellij.openapi.util.Key
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiNamedElement
import com.intellij.psi.ResolveState
import com.intellij.psi.impl.source.PsiFileImpl
import com.intellij.psi.scope.PsiScopeProcessor
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.psi.stubs.StubIndex
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.psi.util.PsiTreeUtil.findChildrenOfType
import net.masterthought.dlanguage.index.DModuleIndex
import net.masterthought.dlanguage.psi.*
import net.masterthought.dlanguage.psi.impl.DLangIdentifierImpl
import net.masterthought.dlanguage.psi.interfaces.DNamedElement
import net.masterthought.dlanguage.psi.interfaces.Declaration
import net.masterthought.dlanguage.psi.interfaces.VariableDeclaration
import net.masterthought.dlanguage.psi.references.DReference
import net.masterthought.dlanguage.stubs.index.DTopLevelDeclarationIndex

/**
 * Created by francis on 5/12/17.
 */
object DResolveUtil {

    class Import(val import: DLangImport) {

        fun isScoped(): Boolean {
            val parent: DLangImportList = import.parent as DLangImportList
            return parent.importBindList != null
        }

        fun getScopedSymbolNames(): MutableSet<String> {
            val names: MutableSet<String> = mutableSetOf<String>()
            val dLangImportList = import.parent as DLangImportList
            val importBinds = findChildrenOfType(dLangImportList.importBindList, DLangImportBind::class.java)
            for (importBind in importBinds) {
                names.add(importBind.identifierList[0].name)
            }
            return names
        }

        /**
         * in theory a module could be from more than one file
         */
        fun getFiles(): MutableSet<DLangFile> {
            val files: MutableSet<DLangFile> = mutableSetOf()
            val reference: DReference = DUtil.getEndOfIdentifierList(import.moduleFullyQualifiedName).reference as DReference
            for (resolveResult in reference.multiResolve(false)) {
                files.add(resolveResult.element?.containingFile as DLangFile)
            }
            return files
        }

    }

    fun getImportedFromLocation(location: DLangIdentifier): MutableSet<Import> {
        class ImportScopeProcessor(val start: DLangIdentifier) : PsiScopeProcessor {
            val result: MutableSet<Import> = mutableSetOf()

            override fun handleEvent(event: PsiScopeProcessor.Event, associated: Any?) {
                return
            }

            override fun <T : Any?> getHint(hintKey: Key<T>): T? {
                return null
            }

            override fun execute(element: PsiElement, state: ResolveState): Boolean {
                if (element is DLangImport) {
                    result.add(Import(element))
                }
                if (element is DLangImportDeclaration) {
                    for (dLangImport in findChildrenOfType(element, DLangImport::class.java)) {
                        result.add(Import(dLangImport))
                    }
                }
                return true
            }
        }

        val importScopeProcessor = ImportScopeProcessor(location)
        PsiTreeUtil.treeWalkUp(importScopeProcessor, location, location.containingFile.containingDirectory, ResolveState.initial())
        return importScopeProcessor.result
    }

    private fun stubTreeCheck(location: DLangIdentifier) {
        if ((location as DLangIdentifierImpl).greenStub == null) {
            (location.containingFile as PsiFileImpl).calcStubTree()
            assert((location.containingFile as DLangFile).greenStub != null)
        }
    }

    fun resolveThroughLocalScope(identifier: DLangIdentifier): DNamedElement? {
        class LocalScopeResolveProcessor(val start: DLangIdentifier) : PsiScopeProcessor {
            var result: DNamedElement? = null

            override fun handleEvent(event: PsiScopeProcessor.Event, associated: Any?) {
                return
            }

            override fun <T : Any?> getHint(hintKey: Key<T>): T? {
                return null
            }

            override fun execute(element: PsiElement, state: ResolveState): Boolean {
                if (element is DNamedElement) {
                    if (element is VariableDeclaration) {
                        if (!element.actuallyIsDeclaration()) {
                            return true
                        }
                    }
                    if (element.name.equals(start.name)) {
                        result = element
                        return false
                    }
                }
                return true
            }
        }

        val resolveProcessor = LocalScopeResolveProcessor(identifier)
        PsiTreeUtil.treeWalkUp(resolveProcessor, identifier, identifier.containingFile.containingDirectory, ResolveState.initial())
        return resolveProcessor.result
    }


    fun inModuleName(start: DLangIdentifier): Boolean {
        val moduleFullyQualifiedName = PsiTreeUtil.getTopmostParentOfType(start, DLangModuleFullyQualifiedName::class.java)
        return moduleFullyQualifiedName != null
    }

    /**
     * Finds name definition across all Haskell files in the project. All
     * definitions are found when name is null.
     */
    fun findDefinitionNode(project: Project, e: DLangIdentifier): List<PsiNamedElement> {
        val inModuleName: Boolean = inModuleName(e)
        if (inModuleName) {
            val name = PsiTreeUtil.getTopmostParentOfType(e, DLangModuleFullyQualifiedName::class.java)!!.text
            val files = DModuleIndex.getFilesByModuleName(project, name, GlobalSearchScope.allScope(project))
            return files.toList()
        }

        val resolveThroughLocalScope = resolveThroughLocalScope(e)
        if (resolveThroughLocalScope != null)
            return listOf(resolveThroughLocalScope)
        val result: MutableList<Declaration> = mutableListOf()
        for (import in getImportedFromLocation(e)) {
            for (file in import.getFiles()) {
                result.addAll(StubIndex.getInstance().get(DTopLevelDeclarationIndex.KEY, e.name, e.project, GlobalSearchScope.fileScope(file)))
            }
        }
        return result


    }

//        // Guess where the name could be defined by lookup up potential modules.
//        val potentialModules = if (e == null)
//            Collections.emptySet<String>();
//        else
//            DPsiUtil.parseImports(e.containingFile)
//
//        val result = ContainerUtil.newArrayList<PsiNamedElement>()
//        val psiFile = if (e == null) null else e.containingFile.originalFile
//        // find definition in current file
//        if (psiFile is DLangFile) {
//            findDefinitionNode(psiFile as DLangFile?, name, e, result)
//        }
//        // find definition in imported files
//        for (potentialModule in potentialModules) {
//            val files = DModuleIndex.getFilesByModuleName(project, potentialModule, GlobalSearchScope.allScope(project))
//            for (f in files) {
//                val returnAllReferences = name == null
//                val inLocalModule = f != null && f == psiFile
//                val inImportedModule = f != null && potentialModules.contains(f.moduleName)
//                if (returnAllReferences || inLocalModule || inImportedModule) {
//                    findDefinitionNode(f, name, e, result)
//                }
//            }
//        }
//        return result
}


/**
 * finds definition(s) of functions/class/template
 * @param file the file to search for definitions in
 * *
 * @param name the name of the function/class/template to resolve
 * *
 * @param e the usage of the defined function/class/template etc.
 * *
 * @param result the results are added to the is arraylist
 */
//    fun findDefinitionNode(file: DLangFile?, name: String?, e: PsiNamedElement?, result: MutableList<PsiNamedElement>) {
//        if (file == null) return
//        // start with empty list of potential named elements
//        val declarationElements = ArrayList<DNamedElement>()
//
//        if (e is DLangIdentifier) {
//
//            val declarations = ArrayList<Declaration>()
//            //            final Collection<DNamedElement> elements = StubIndex.getElements(DAllNameIndex.KEY, e.getName(), e.getProject(), GlobalSearchScope.fileScope(file), DNamedElement.class);
//            val elements = StubIndex.getInstance().get(DAllNameIndex.KEY, e.name, e.project, GlobalSearchScope.fileScope(file))
//            for (element in elements) {
//                if (element is Declaration) {
//                    declarations.add(element)
//                }
//            }
//
//            for (candidateDeclaration in declarations) {
//                if (candidateDeclaration is DLangAutoDeclarationY) {
//                    if (candidateDeclaration.actuallyIsDeclaration()) {
//                        declarationElements.add(candidateDeclaration)
//                    }
//                    continue
//                }
//                declarationElements.add(candidateDeclaration)
//            }
//        }
//
//        var resolvingConstructor = false
//
//        var parent: PsiElement? = e!!.parent
//        while (true) {
//            if (parent == null)
//                break
//            if (parent is DLangNewExpression || parent is DLangNewExpressionWithArgs)
//                resolvingConstructor = true
//            parent = parent.parent
//        }
//
//        // check the list of potential named elements for a match on name
//        for (namedElement in declarationElements) {
//            //non void initializer
//            if (resolvingConstructor) {
//                if (namedElement is DLangConstructor) {
//                    result.add(namedElement)
//                }
//            } else if (name == null || name == namedElement.name && e != namedElement && namedElement !is DLangConstructor) {
//                result.add(namedElement)
//            }
//        }
//    }

/**
 * Finds a name definition inside a D file. All definitions are found when name
 * is null.
 */
//    fun findDefinitionNodes(dLangFile: DLangFile?, name: String?): MutableList<PsiNamedElement> {
//        val ret = ContainerUtil.newArrayList<PsiNamedElement>()
//        findDefinitionNode(dLangFile, name, null, ret)
//        return ret
//    }

/**
 * Finds name definition across all D files in the project. All
 * definitions are found when name is null.
 */
//    fun findDefinitionNodes(project: Project): List<PsiNamedElement> {
//        return findDefinitionNode(project, null)
//    }

/**
 * Finds name definitions that are within the scope of a file, including imports (to some degree).
 */
//    fun findDefinitionNodes(psiFile: DLangFile): List<PsiNamedElement> {
//        val result = findDefinitionNodes(psiFile, null)
//        result.addAll(findDefinitionNode(psiFile.project, null))
//        return result
//    }

